java/arcs/core/data/expression/Evaluator.kt

```diff
@@ -53,6 +53,110 @@ class ExpressionEvaluator(
         currentScope
 
     override fun <T> visit(expr: Expression.ObjectLiteralExpression<T>): Any = expr.value as Any
+
+    override fun <E, T> visit(expr: Expression.FromExpression<E, T>): Any {
+        var sequence = requireNotNull(currentScope.lookup<Any>(expr.source)) {
+            "${expr.source} is null in current scope."
+        }
+
+        if (sequence is List<*>) {
+            sequence = sequence.asSequence()
+        }
+
+        require(sequence is Sequence<*>) {
+            "${expr.source} of type ${sequence::class} cannot be converted to a Sequence"
+        }
+
+        return sequence.map { value ->
+            currentScope.set(expr.iterationVar, value as Any)
+            value
+        }
+    }
+
+    override fun <T> visit(expr: Expression.WhereExpression<T>): Any {
+        return (expr.qualifier.accept(this) as Sequence<T>).filter {
+            expr.expr.accept(this) == true
+        }
+    }
+
+    override fun <E, T> visit(expr: Expression.SelectExpression<E, T>): Any {
+        return (expr.qualifier.accept(this) as Sequence<E>).map {
+            expr.expr.accept(this) as T
+        }
+    }
+
+    override fun <T> visit(expr: Expression.NewExpression<T>): Any {
+        val newScope = scopeCreator(expr.schemaName.firstOrNull() ?: "")
+        expr.fields.forEach { (fieldName, fieldExpr) ->
+            newScope.set(fieldName, fieldExpr.accept(this))
+        }
+        return newScope
+    }
+
+    override fun <T> visit(expr: Expression.FunctionExpression<T>): Any {
+        val arguments = expr.arguments.map { it.accept(this) }.toList()
+        return expr.function.invoke(arguments)
+    }
+}
+
+private fun <T> toSequence(value: Any?) = when (value) {
+    null -> emptySequence<T>()
+    is Sequence<*> -> value as Sequence<T>
+    is Collection<*> -> value.asSequence() as Sequence<T>
+    else -> sequenceOf<T>(value as T)
+}
+
+/** Global functions are invoked by [FunctionExpression] during evaluation. */
+sealed class GlobalFunction(val name: String) {
+
+    /** Functions accept a varargs list of [Sequence] and return any type. */
+    abstract fun invoke(args: List<Any>): Any
+
+    /** Performs [Sequence.union] of two [Sequence]s. */
+    object Union : GlobalFunction("union") {
```

Comment:
Do you have strong preference wrt casing? CamelCase may be better if we ever have functions that have multi-word names. Though you could do under_scopes there. 

I would argue for using the same format here and in the parser, just to limit weirdness. WDYT?

Reply:
You mean you want the function to be case sensitive "Union"? That's more a C# rule though, because in Java/Kotlin, it would begin with lowercase char?

I don't have a preference, whatever is preferred.